Introduce "cross-origin-isolated" permission

yutakahirano opened the pull request:
Introduce "cross-origin-isolated" permission to allow a document to
control whether nested documents can access features that require
securer context, even when COOP+COEP are enabled.

Fixes #5435.

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Firefox
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/24600
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1115379
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1658677
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=215407

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5752/infrastructure.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/5752/631a32e...cb23065/infrastructure.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">diff</a> )
<a href="https://whatpr.org/html/5752/webappapis.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/5752/631a32e...cb23065/webappapis.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">diff</a> )
<a href="https://whatpr.org/html/5752/window-object.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">/window-object.html</a>  ( <a href="https://whatpr.org/html/5752/631a32e...cb23065/window-object.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">diff</a> )
<a href="https://whatpr.org/html/5752/workers.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">/workers.html</a>  ( <a href="https://whatpr.org/html/5752/631a32e...cb23065/workers.html" title="Last updated on Aug 12, 2020, 4:01 PM UTC (cb23065)">diff</a> )

yutakahirano commented:
@domenic @annevk Can you PTAL?

For (nested) documents We use [#allowed-to-use](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#allowed-to-use).
For dedicated workers the permission is inherited from the ancestor document.

Shared and service workers don't inherit the policy. This comes from the fact that we don't inherit COEP for these workers, but web developers who wants to use APIs requireing securer context to sniff process memory shared with the parent document can use service worker to do that, assuming the worker and the frame are hosted by the same process (this is not guaranteed but they often are). Do you think it's problematic? I thought about this for a while but couldn't find a good solution.

annevk commented:
Well, per the current model a shared/service worker that is cross-origin isolated would not be put in the same process with something that is not cross-origin isolated. It's not really clear how you delegate to shared/service workers though and whether shared workers need to be isolated the way they are for secure contexts. There's quite a few gotchas it seems with using permission delegation here (and perhaps in general when workers are important).

yutakahirano commented on source:
```diff
@@ -92550,8 +92554,27 @@ document.body.appendChild(frame)</code></pre>
   <p>The <dfn data-x="dom-crossOriginIsolated"><code>crossOriginIsolated</code></dfn> getter steps
-  are to return the <span>surrounding agent</span>'s <span>agent cluster</span>'s <span>cross-origin
-  isolated</span>.</p>
+  are to return true if all of the following conditions are true, and false otherwise.</p>
```
Done.

yutakahirano commented:
@domenic, do you have any opinions?

domenic commented:
Sorry about not responding earlier. I can be more responsive on this going forward, now that I understand it more fully.

I think shared/service workers should be treated as "roots". Their cross-origin isolated status should only be controlled by themselves and the headers they set.

For dedicated workers, automatically inheriting from their parent seems fine, since they must be same-origin to their parent. (Or they can be data: URLs, but data: URLs also inherit this sort of thing, generally.)

---

On this PR generally, it seems like we will need to refactor things so that the `crossOriginIsolated` getter steps can be called from elsewhere. In particular I wonder if it's confusing to have separate "cross-origin isolated" (the boolean set at creation time) and "effectively cross-origin isolated" (what the getter steps return). Plus the "cross-origin isolated permission" boolean. I'm not sure what the best architecture is though, and perhaps it will be modified given my above comments...

yutakahirano commented:
How about the following?

-----

### [Environment settings object](https://html.spec.whatwg.org/multipage/webappapis.html#environment-settings-objects)

An environment settings object is an environment that additionally specifies algorithms for:
...

__An effective cross-origin isolated__: A boolean represents whether it is allowed to use APIs that requires cross-origin isolation.

### [Script settings for Window objects](https://html.spec.whatwg.org/multipage/window-object.html#script-settings-for-window-objects)

The effective cross-origin isolated: Return the logical conjunction of the surrounding agent's agent cluster's cross-origin isolated and whether its associated Document is allowed to use the "cross-origin-isolated" permission.

### [Script settings for workers](https://html.spec.whatwg.org/multipage/workers.html#script-settings-for-workers)

The effective cross-origin isolated: Return worker global scope's effective cross-origin isolated.

(Note: worker global scope's effective cross-origin isolated is defined similarly to the current change).

### [The WindowOrWorkerGlobalScope mixin](https://html.spec.whatwg.org/multipage/webappapis.html#windoworworkerglobalscope-mixin)

self.crossOriginIsolated returns the relevant settings object's effective cross-origin isolated.

domenic commented:
That sounds pretty good to me! Two notes:

* Perhaps better names would be something like "COOP+COEP headers set" (for Window's current "cross-origin isolated") and "cross-origin isolated" (for the settings object field, which we call above "effective cross-origin isolated"). I would like to avoid any object having cross-origin isolated = true if `thatObject.crossOriginIsolated` returns false.

* As usual, adding new fields to settings objects requires modifying https://w3c.github.io/ServiceWorker/#run-service-worker-algorithm and https://drafts.css-houdini.org/worklets/#script-settings-for-worklets

yutakahirano commented:
Updated.

I like "agent cluster's cross-origin isolated" more than "COOP+COEP headers set". We use the value for important decisions (e.g., document.domain behavior) and while it may not be exposed via crossOriginIsolated I think "agent cluster's cross-origin isolated" is more important, and deserves a simpler name.

I will create PRs for service workers and worklets.

domenic commented:
> I like "agent cluster's cross-origin isolated" more than "COOP+COEP headers set". We use the value for important decisions (e.g., document.domain behavior)

Interesting. And I guess we don't want to turn document.domain back on if the permission isn't delegated. OK, that makes sense then.

I wonder then about the name "effective cross-origin isolated". Maybe "allowed to use cross-origin isolated features"? Any bikeshedding is welcome. (But there's no need to update the PR right now; we can wait until everyone is settled to go through more changes.)

> I will create PRs for service workers and worklets.

Actually, since I already have PRs out for related changes, how about you just add comments to https://github.com/w3c/ServiceWorker/pull/1527 and https://github.com/w3c/css-houdini-drafts/pull/998 saying what the algorithms should be, and I can edit my PRs to include them?

domenic commented on source:
```diff
@@ -79107,7 +79111,15 @@ interface <dfn>BarProp</dfn> {
+     cluster</span>'s <span>cross-origin isolated</span> and whether <var>window</var>'s <span
+     data-x="concept-document-window">associated <code>Document</code></span> is <span>allowed to
+     use</span> the "<code
+     data-x="cross-origin-isolated-permission">cross-origin-isolated</code>".</p></dd>
```
```suggestion
     data-x="cross-origin-isolated-permission">cross-origin-isolated</code>" feature.</p></dd>
```

yutakahirano replied:
Fixed.

domenic commented on source:
```diff
@@ -4197,6 +4197,10 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
     <li>"<dfn data-x="autoplay-feature"><code data-x="">autoplay</code></dfn>", which has a <span
     data-x="concept-default-allowlist">default allowlist</span> of <code
     data-x="">'self'</code>.</li>
+    <li>"<dfn data-x="cross-origin-isolated-permission"><code
```
-permission should probably be -feature?

yutakahirano replied:
Fixed.

domenic commented on source:
```diff
@@ -87565,6 +87577,12 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+   <dt>An <dfn data-x="concept-settings-object-effective-cross-origin-isolated" data-export=""
+   data-dfn-for="environment settings object">effective cross-origin isolated</dfn> boolean</dt>
+
+   <dd><p>A boolean representing whether it is allowed to use APIs that requires cross-origin
```
```suggestion
   <dd><p>A boolean representing whether it is allowed to use APIs that require cross-origin
```

domenic commented on source:
```diff
@@ -92513,12 +92531,11 @@ interface mixin <dfn>WindowOrWorkerGlobalScope</dfn> {
-   data-x="">Cross-Origin-Embedder-Policy</code>` HTTP response headers and determines whether
-   <code>SharedArrayBuffer</code> can be used with <code data-x="">postMessage()</code>
-   APIs.</p></dd>
+   <dd><p>Returns whether it is allowed to use APIs that require cross-origin isolattion.
```
```suggestion
   <dd><p>Returns whether it is allowed to use APIs that require cross-origin isolation.
```

I wonder what "it" should be here. "Script running in this global", maybe?

domenic commented on source:
```diff
@@ -92513,12 +92531,11 @@ interface mixin <dfn>WindowOrWorkerGlobalScope</dfn> {
+   This depends on the `<code data-x="">Cross-Origin-Opener-Policy</code>` and `<code
+   data-x="">Cross-Origin-Embedder-Policy</code>` HTTP response headers and the
+   "<code data-x="cross-origin-isolated-permission">cross-origin-isolated</code>"
+   permission.</p></dd>
```
```suggestion
   feature.</p></dd>
```

domenic commented on source:
```diff
@@ -99410,6 +99444,13 @@ interface <dfn>SharedWorkerGlobalScope</dfn> : <span>WorkerGlobalScope</span> {
+     data-x="concept-url-scheme">scheme</span> is "<code data-x="">data</code>", then set
+     <var>worker global scope</var>'s <span
+     data-x="concept-WorkerGlobalScope-effective-cross-origin-isolated">effective cross-origin
+     isolated</span> to false.
```
What is the reasoning for this? data: URL workers have their own separate agent clusters, so I would think they should be allowed to use cross-origin isolated features...

yutakahirano replied:
This is because "data:" origins are opaque. We need these worker-specific logic because permission policy doesn't work with workers (https://github.com/w3c/webappsec-permissions-policy/issues/207).

domenic replied:
Right, they're opaque, so they should be treated as cross-origin isolated, right? So this should be set to true, if anything...

yutakahirano replied:
I'm talking about hte "cross-origin-isolated" feature. The worker is cross origin compared to its owner, and it is not able to set permission policies for data: workers so the default policy ("self") is applied, which means effective cross-origin isolated is false.

yutakahirano replied:
Also, according to https://html.spec.whatwg.org/multipage/webappapis.html#obtaining-a-worker/worklet-agent and https://html.spec.whatwg.org/multipage/webappapis.html#obtain-a-dedicated/shared-worker-agent a dedicated worker's agent cluster is its owner's.

domenic replied:
Sorry, I was confused with `data:` URL iframes, which are their own agent cluster. OK, this is probably fine... I'll do a more full re-review.

domenic replied:
Hmm, upon re-reviewing I am still confused why we want to prevent data: URL workers from using cross-origin isolated features. I know they have a different origin. However, for other important things, we inherit from the parent, even though they are cross-origin. For example, for embedder policy.

yutakahirano replied:
Yeah, we can choose either way. Ultimately that will be resolved in w3c/webappsec-permissions-policy#207 but to be honest I don't want to be blocked by the issue. I'm fine with either options, but starting with effective cross-origin isolated: false and changing the value to true if necessary (when w3c/webappsec-permissions-policy#207 is resolved) seems safer than the other option from the compatibility perspective.

Please note that for COEP we didn't have options. As the owner and the worker share the same agent cluster we needed to inherit the value when the owner had COEP: require-corp.

@annevk  do you have opinions?

domenic replied:
Got it. I understand now, and agree this is a good conservative default. I will push a commit adding a note about that reasoning.

yutakahirano commented:
Thank you the review! Somehow "commit suggestion" didn't work so I addressed the comments manually.

Reg feature and permission: I was under the impression that spec authors were replacing "feature" with "permission" (https://github.com/w3c/webappsec-permissions-policy/issues/359), but I'm happy to accept your suggestion in this change.

domenic commented:
@clelland are Permissions Policy things called "permissions" now, or still "features"? The spec still talks about features ([example](https://w3c.github.io/webappsec-permissions-policy/#algo-define-inherited-policy)).

domenic commented on source:
```diff
@@ -99370,6 +99393,17 @@ interface <dfn>SharedWorkerGlobalScope</dfn> : <span>WorkerGlobalScope</span> {
+   <li><p>Set <var>worker global scope</var>'s <span
+   data-x="concept-WorkerGlobalScope-effective-cross-origin-isolated">effective cross-origin
+   isolated</span> to <var>agent</var>'s <span>agent cluster</span>'s <span>cross-origin
+   isolated</span>.</p></li>
```
This seems problematic because we only set the agent's agent cluster's cross-origin isolated in step 14.7. Probably this needs to move down there?

yutakahirano replied:
Good catch, thank you. Fixed.

domenic commented:
In IRC @annevk suggested "cross-origin isolated capability" as the best name for what this PR calls "effective cross-origin isolated". I am happy with that as the final name. I will push a commit renaming in that direction.

yutakahirano commented:
Thank you! I filed a bug for each implementation.
